### ISSUE TITLE
multi label MMOD slow training

### ISSUE THREAD

**Issue by AndreaPisoni** [davisking/dlib#840]
I am trying to train a multi-labeled mmod detector for a few classes. I have noticed the training is about 10x slower than when I used to train single-class objects (same dataset - just no labels) on dlib 19.4.  Is this expected? I thought the speed of the detector would be similar... 

I have ran nvidia-smi and noticed that the GPU is barely used (10-30%) and the data is fed from the CPU on a single thread. I thought there would be a "num_threads" option on dnn_trainer, but I can't see it... any way to feed data to the GPU from multiple threads? Maybe this would help speeding the training up...

**Comment by davisking** [comment]
There isn't any speed difference between the previous version and the new version. The code is basically identical.  It's on you to load mini-batches fast enough.  If you need to use multiple threads to do that then do so and pass them to the trainer via train_one_step().  

You also might just need to wait a minute.  Some datasets generate a high number of false alarms which cause the MMOD loss to be computed slowly, which happens on the CPU.  After a little bit that will go away and the GPU will be fully loaded.  Maybe that's all that's happening.

**Comment by AndreaPisoni** [comment]
> You also might just need to wait a minute. Some datasets generate a high number of false alarms which cause the MMOD loss to be computed slowly, which happens on the CPU. After a little bit that will go away and the GPU will be fully loaded. Maybe that's all that's happening.

Is it really just that? I let the code run for one hour... same dataset, same batch size, I get about 2000 steps in 5 minutes with the dlib 19.4 single label MMOD detector, and about 150 steps with dlib 19.7 multi-label clas... this does not look right at all...

**Comment by davisking** [comment]
It's still not loading the GPU after an hour?

**Comment by AndreaPisoni** [comment]
Yes, I am checking nvidia-smi every 0.5s and the GPU is 0% - 15% most of the time. I am saving the sync file every 5 minutes and in 5 minutes I get about 20 steps in total... (I raised the batch size to 80 from 25 to try out whether it would be faster...) I have and Nvidia 1080 and an intel 6700k so I would expect the performances to be better... but especially, I get totally different performances with the dlib 19.4 single class detector...

**Comment by AndreaPisoni** [comment]
Extra info: the dataset fits in RAM completely, so I would not think it's an IO issue...

**Comment by davisking** [comment]
How many classes?  

Put a print statement here https://github.com/davisking/dlib/blob/master/dlib/dnn/loss.h#L744 and see how big dets is.  Maybe it's huge.

**Comment by AndreaPisoni** [comment]
I got 22 classes. 

Bingo! Dets seems extremely large, more than 450k elements. Why is that? Is it because I got too many aspect ratios, so the (num classes * num aspect ratios) number explodes?

**Comment by davisking** [comment]
Yeah that’s what’s happening. I have code in there that limits part of the detection processing to deal with this.  You always get a lot of detections in the early iterations and it’s going to be linear in the classes and aspect ratios. 

Really we should just throw away a lot of the detections. Try timing the code so see what part is slow and discarding detections. There is already code that does this. Either the discard amount is not right or the slowness is due to the little bit of code before the discard.

**Comment by AndreaPisoni** [comment]
I  tried resizing to one single aspect ratio (of course for now this defeats the purpose) and the speed now is on par with what i was seeing in 19.4. Dets was 2k in the first few iteration and then rapidly went down to 5-10. 

I am thinking maybe I should run k-means and get the "10 most signficant aspect ratios" and resize my data to those, YOLO and other single shot detector use similar techniques to choose their "anchor boxes" that will be used as detection candidates.

What do you think? Is there anything better we can do? Is your idea to try to trim the detectors more solid?

**Comment by davisking** [comment]
It already clusters the aspect ratios. I would trim the detections. I thought the code in there was already trimming it plenty but I guess not. Trimming is definitely the right thing.

**Comment by davisking** [comment]
As an example, try something like this:

```
--- a/dlib/dnn/loss.h   Wed Sep 20 21:10:59 2017 -0400
+++ b/dlib/dnn/loss.h   Sun Sep 24 14:37:38 2017 -0400
@@ -725,6 +725,7 @@
             DLIB_CASSERT(input_tensor.num_samples() == output_tensor.num_samples());
             DLIB_CASSERT(output_tensor.k() == (long)options.detector_windows.size());
 
+            double det_thresh_speed_adjust = 0;
 
 
             // we will scale the loss so that it doesn't get really huge
@@ -741,9 +742,15 @@
             std::vector<intermediate_detection> dets;
             for (long i = 0; i < output_tensor.num_samples(); ++i)
             {
-                tensor_to_dets(input_tensor, output_tensor, i, dets, -options.loss_per_false_alarm, sub);
+                tensor_to_dets(input_tensor, output_tensor, i, dets, -options.loss_per_false_alarm + det_thresh_speed_adjust, sub);
 
                 const unsigned long max_num_dets = 50 + truth->size()*5;
+                // Prevent calls to tensor_to_dets() from running for a really long time
+                // due to the production of an obscene number of detections.
+                const unsigned long max_num_initial_dets = max_num_dets*100;
+                if (dets.size() >= max_num_initial_dets)
+                    det_thresh_speed_adjust = std::max(det_thresh_speed_adjust,dets[max_num_initial_dets].detection_confidence + options.loss_per_false_alarm);
+
```

Does that kind of construction deal with the problem you are having?

**Comment by AndreaPisoni** [comment]
Thanks Davis. The above code makes the training about 2x faster, but still about ~6x slower than clustering the data to have only 10 aspect ratios.

**Comment by davisking** [comment]
How many aspect ratios did you have before?

**Comment by AndreaPisoni** [comment]
I did not run any regularization on the data aspect ratio before, so I guess my boxes will have very similar aspect ratio among the same class, but be slightly different due to having humans annotating the images.

**Comment by davisking** [comment]
That shouldn't matter, there is an aspect ratio clustering already done inside the mmod_options object.  The only possible problem would be if you had some extreme outliers since they would be assigned their own cluster.  So if you filtered out outliers then it would be better, otherwise it shouldn't matter.
